### PR TITLE
Add Velocity volume output for compressible flow solvers

### PR DESCRIPTION
**Pull request by MastermindScope**
## Disclaimer
This is my first ever contribution to any GitHub project. Please give feedback on what I can Improve if I haven't done something that is expected of me.

## Proposed Changes
Added Velocity output, since I am interested in that and adding it only in the config just gets ignored and no output happens
 

## PR Checklist

- [X] I am submitting my contribution to the develop branch.
- [ ] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [ ] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

### CODE REVIEW

**Review by pcarruscag — COMMENTED** [review]
Welcome to SU2. You need to add these variables to the "PRIMITIVE" group, there is a comment `// Primitive variables`.
Look also for `SetVolumeOutputValue("PRESSURE"` because for each variable you need to add it and then set its value.
Please make the same change in CNEMOCompOutput.cpp
And add yourself to AUTHORS.md

**Comment by MastermindScope** [comment]
Hi, thank you for the warm welcome. I added the changes and hope they are sufficient.

File: SU2_CFD/src/output/CNEMOCompOutput.cpp [hunk 1]
```diff
@@ -242,10 +242,15 @@ void CNEMOCompOutput::SetVolumeOutputFields(CConfig *config){
...
   AddVolumeOutput("TEMPERATURE_VE", "Temperature_ve", "PRIMITIVE", "Temperature_ve");
+  AddVolumeOutput("VELOCITY-X", "Velocity_x", "PRIMITIVE", "x-component of the velocity vector");
+  AddVolumeOutput("VELOCITY-Y", "Velocity_y", "PRIMITIVE", "y-component of the velocity vector");
 
   AddVolumeOutput("MACH",        "Mach",                    "PRIMITIVE", "Mach number");
   AddVolumeOutput("PRESSURE_COEFF", "Pressure_Coefficient", "PRIMITIVE", "Pressure coefficient");
 
+  if (nDim == 3)
+    AddVolumeOutput("VELOCITY-Z", "Velocity_z", "PRIMITIVE", "z-component of the velocity vector");
+
```
Comment (pcarruscag) [inline SU2_CFD/src/output/CNEMOCompOutput.cpp:253]:
```suggestion
  if (nDim == 3)
    AddVolumeOutput("VELOCITY-Z", "Velocity_z", "PRIMITIVE", "z-component of the velocity vector");

  AddVolumeOutput("MACH",        "Mach",                    "PRIMITIVE", "Mach number");
  AddVolumeOutput("PRESSURE_COEFF", "Pressure_Coefficient", "PRIMITIVE", "Pressure coefficient");

```

**Review by pcarruscag — COMMENTED** [review]
Thanks, you are still missing the part of the code that sets the output value, AddVolumeOutput only adds the name.
Look for SetVolumeOutputValue in the files you changed, should be very close.

**Comment by MastermindScope** [comment]
I've added SetVolumeOutputValue now. Thank you for being patient and letting me make mistakes.

**Comment by EvertBunschoten** [comment]
Hi! Just out of curiosity, if you're interested in just the velocity and not the momentum, why not just divide the momentum vector by the density during postprocessing?

File: SU2_CFD/src/output/CFlowCompOutput.cpp [hunk 2]
```diff
@@ -310,11 +316,14 @@ void CFlowCompOutput::LoadVolumeData(CConfig *config, CGeometry *geometry, CSolv
...
+  SetVolumeOutputValue("VELOCITY-Y", iPoint, Node_Flow->GetSolution(iPoint, 4));
   if (nDim == 3){
-    SetVolumeOutputValue("MOMENTUM-Z", iPoint, Node_Flow->GetSolution(iPoint, 3));
-    SetVolumeOutputValue("ENERGY",     iPoint, Node_Flow->GetSolution(iPoint, 4));
+    SetVolumeOutputValue("MOMENTUM-Z", iPoint, Node_Flow->GetSolution(iPoint, 5));
+    SetVolumeOutputValue("ENERGY",     iPoint, Node_Flow->GetSolution(iPoint, 6));
+    SetVolumeOutputValue("VELOCITY-Z", iPoint, Node_Flow->GetSolution(iPoint, 7));
   } else {
-    SetVolumeOutputValue("ENERGY",     iPoint, Node_Flow->GetSolution(iPoint, 3));
+    SetVolumeOutputValue("ENERGY",     iPoint, Node_Flow->GetSolution(iPoint, 5));
```
Comment (EvertBunschoten) [inline SU2_CFD/src/output/CFlowCompOutput.cpp:326]:
```suggestion
    SetVolumeOutputValue("ENERGY",     iPoint, Node_Flow->GetSolution(iPoint, 3));
```
Please leave the solution indices of the other output variables as they are. The total energy of the flow is stored in solution index 3 during the solution process, not 5. Now the energy output values are incorrect.

File: SU2_CFD/src/output/CFlowCompOutput.cpp [hunk 3]
```diff
@@ -310,11 +316,14 @@ void CFlowCompOutput::LoadVolumeData(CConfig *config, CGeometry *geometry, CSolv
   SetVolumeOutputValue("DENSITY",    iPoint, Node_Flow->GetSolution(iPoint, 0));
   SetVolumeOutputValue("MOMENTUM-X", iPoint, Node_Flow->GetSolution(iPoint, 1));
   SetVolumeOutputValue("MOMENTUM-Y", iPoint, Node_Flow->GetSolution(iPoint, 2));
+  SetVolumeOutputValue("VELOCITY-X", iPoint, Node_Flow->GetSolution(iPoint, 3));
```
Comment (EvertBunschoten) [inline SU2_CFD/src/output/CFlowCompOutput.cpp:319]:
```suggestion
  SetVolumeOutputValue("VELOCITY-X", iPoint, Node_Flow->GetPrimitive(iPoint, 1));
```
For compressible flows, the velocity is stored in the `Primitive` data, not the `Solution` data. The indices for the x,y, and z velocity components in the `Primitive` data are 1,2, and 3 respectively in the `CEulerVariable` class.

File: SU2_CFD/src/output/CFlowCompOutput.cpp [hunk 4]
```diff
@@ -310,11 +316,14 @@ void CFlowCompOutput::LoadVolumeData(CConfig *config, CGeometry *geometry, CSolv
...
+  SetVolumeOutputValue("VELOCITY-Y", iPoint, Node_Flow->GetSolution(iPoint, 4));
   if (nDim == 3){
-    SetVolumeOutputValue("MOMENTUM-Z", iPoint, Node_Flow->GetSolution(iPoint, 3));
-    SetVolumeOutputValue("ENERGY",     iPoint, Node_Flow->GetSolution(iPoint, 4));
+    SetVolumeOutputValue("MOMENTUM-Z", iPoint, Node_Flow->GetSolution(iPoint, 5));
+    SetVolumeOutputValue("ENERGY",     iPoint, Node_Flow->GetSolution(iPoint, 6));
+    SetVolumeOutputValue("VELOCITY-Z", iPoint, Node_Flow->GetSolution(iPoint, 7));
```
Comment (EvertBunschoten) [inline SU2_CFD/src/output/CFlowCompOutput.cpp:324]:
```suggestion
    SetVolumeOutputValue("VELOCITY-Z", iPoint, Node_Flow->GetPrimitive(iPoint, 3));
```

File: SU2_CFD/src/output/CNEMOCompOutput.cpp [hunk 5]
```diff
@@ -313,13 +317,16 @@ void CNEMOCompOutput::LoadVolumeData(CConfig *config, CGeometry *geometry, CSolv
 
   SetVolumeOutputValue("MOMENTUM-X", iPoint, Node_Flow->GetSolution(iPoint, nSpecies));
   SetVolumeOutputValue("MOMENTUM-Y", iPoint, Node_Flow->GetSolution(iPoint, nSpecies+1));
+  SetVolumeOutputValue("VELOCITY-X", iPoint, Node_Flow->GetSolution(iPoint, nSpecies+2));
```
Comment (EvertBunschoten) [inline SU2_CFD/src/output/CNEMOCompOutput.cpp:320]:
```suggestion
  SetVolumeOutputValue("VELOCITY-X", iPoint, Node_Flow->GetPrimitive(iPoint, nSpecies+2));
```
Similar as to the `CEulerVariable` class, for the `CNEMOEulerVariable` class, the velocity is stored in the `Primitive` vector, not the `Solution` vector.

File: SU2_CFD/src/output/CNEMOCompOutput.cpp [hunk 6]
```diff
@@ -313,13 +317,16 @@ void CNEMOCompOutput::LoadVolumeData(CConfig *config, CGeometry *geometry, CSolv
...
-    SetVolumeOutputValue("ENERGY",     iPoint, Node_Flow->GetSolution(iPoint, nSpecies+3));
-    SetVolumeOutputValue("ENERGY_VE",  iPoint, Node_Flow->GetSolution(iPoint, nSpecies+4));
+    SetVolumeOutputValue("MOMENTUM-Z", iPoint, Node_Flow->GetSolution(iPoint, nSpecies+4));
+    SetVolumeOutputValue("ENERGY",     iPoint, Node_Flow->GetSolution(iPoint, nSpecies+5));
+    SetVolumeOutputValue("ENERGY_VE",  iPoint, Node_Flow->GetSolution(iPoint, nSpecies+6));
+    SetVolumeOutputValue("VELOCITY-Z", iPoint, Node_Flow->GetSolution(iPoint, nSpecies+7));
   } else {
-    SetVolumeOutputValue("ENERGY",     iPoint, Node_Flow->GetSolution(iPoint, nSpecies+2));
-    SetVolumeOutputValue("ENERGY_VE",  iPoint, Node_Flow->GetSolution(iPoint, nSpecies+3));
+    SetVolumeOutputValue("ENERGY",     iPoint, Node_Flow->GetSolution(iPoint, nSpecies+4));
```
Comment (EvertBunschoten) [inline SU2_CFD/src/output/CNEMOCompOutput.cpp:328]:
```suggestion
    SetVolumeOutputValue("ENERGY",     iPoint, Node_Flow->GetSolution(iPoint, nSpecies+2));
```
Again mind the solution indices.

**Review by EvertBunschoten — CHANGES_REQUESTED** [review]
Make sure you retrieve the velocity from the `Primitive` vector stored in the `CVariable` classes, not the `Solution` vector.

File: SU2_CFD/src/output/CNEMOCompOutput.cpp (hunk 5 above)
Comment (WallyMaier) [reply SU2_CFD/src/output/CNEMOCompOutput.cpp:320]: ^^^
Additionally, the x-velocity will be at index "nSpecies", not "nSpecies+2"

**Comment by MastermindScope** [comment]
> why not just divide the momentum vector by the density during postprocessing?

There are several reasons why:
1. I thought of this, but it looks patchy somehow
2. It makes life easier for people who want to use this software but have little experience (like me)
3. I want to start contributing to this project and this was one of the things I noticed and thought I could implement myself

File: SU2_CFD/src/output/CFlowCompOutput.cpp (hunk 3 above)
Comment (MastermindScope) [reply SU2_CFD/src/output/CFlowCompOutput.cpp]: This is implemented as well

File: SU2_CFD/src/output/CFlowCompOutput.cpp (hunk 4 above)
Comment (MastermindScope) [reply SU2_CFD/src/output/CFlowCompOutput.cpp]: Did this

File: SU2_CFD/src/output/CNEMOCompOutput.cpp (hunk 5 above)
Comment (MastermindScope) [reply SU2_CFD/src/output/CNEMOCompOutput.cpp]: I have done this now

File: SU2_CFD/src/output/CNEMOCompOutput.cpp (hunk 6 above)
Comment (MastermindScope) [reply SU2_CFD/src/output/CNEMOCompOutput.cpp]: Thank you for reviewing my code.

File: SU2_CFD/src/output/CNEMOCompOutput.cpp [hunk 7]
```diff
@@ -317,16 +317,16 @@ void CNEMOCompOutput::LoadVolumeData(CConfig *config, CGeometry *geometry, CSolv
 
   SetVolumeOutputValue("MOMENTUM-X", iPoint, Node_Flow->GetSolution(iPoint, nSpecies));
   SetVolumeOutputValue("MOMENTUM-Y", iPoint, Node_Flow->GetSolution(iPoint, nSpecies+1));
-  SetVolumeOutputValue("VELOCITY-X", iPoint, Node_Flow->GetSolution(iPoint, nSpecies+2));
-  SetVolumeOutputValue("VELOCITY-Y", iPoint, Node_Flow->GetSolution(iPoint, nSpecies+3));
+  SetVolumeOutputValue("VELOCITY-X", iPoint, Node_Flow->GetPrimitive(iPoint, nSpecies));
```
Comment (EvertBunschoten) [inline SU2_CFD/src/output/CNEMOCompOutput.cpp:320]: Could you please verify whether the current implementation actually returns the correct velocity values before you commit? With the current implementation, the velocity vector is inconsistent with the momentum vector. The primitive indices for x-, y-, and z-velocity should be `nSpecies + 2`, `nSpecies + 3`, and `nSpecies + 4` respectively for the NEMO output.

**Review by EvertBunschoten — APPROVED** [review]

File: SU2_CFD/src/output/CNEMOCompOutput.cpp [hunk 8]
```diff
@@ -313,10 +317,13 @@ void CNEMOCompOutput::LoadVolumeData(CConfig *config, CGeometry *geometry, CSolv
 
   SetVolumeOutputValue("MOMENTUM-X", iPoint, Node_Flow->GetSolution(iPoint, nSpecies));
   SetVolumeOutputValue("MOMENTUM-Y", iPoint, Node_Flow->GetSolution(iPoint, nSpecies+1));
+  SetVolumeOutputValue("VELOCITY-X", iPoint, Node_Flow->GetPrimitive(iPoint, nSpecies));
+  SetVolumeOutputValue("VELOCITY-Y", iPoint, Node_Flow->GetPrimitive(iPoint, nSpecies+1));
   if (nDim == 3){
     SetVolumeOutputValue("MOMENTUM-Z", iPoint, Node_Flow->GetSolution(iPoint, nSpecies+2));
     SetVolumeOutputValue("ENERGY",     iPoint, Node_Flow->GetSolution(iPoint, nSpecies+3));
     SetVolumeOutputValue("ENERGY_VE",  iPoint, Node_Flow->GetSolution(iPoint, nSpecies+4));
+    SetVolumeOutputValue("VELOCITY-Z", iPoint, Node_Flow->GetPrimitive(iPoint, nSpecies+2));
```
Comment (pcarruscag) [inline SU2_CFD/src/output/CNEMOCompOutput.cpp:326]: My other comments is also applicable here. Use GetVelocity and be consistent with the order in which you added the variables

Comment (pcarruscag) [reply SU2_CFD/src/output/CNEMOCompOutput.cpp:326]: Fixed it and pushed to develop

**Pull request merged**